### PR TITLE
Tensorboard logging and EOS weights

### PR DESCRIPTION
**Pull request by liaojh1998**
This branch is used to test out the new changes (RNN/Transformers).
- Added tensorboard logging for all weights.
- Added EOS weights for loss to focus on stop tokens better.

The code is untested as my code is currently getting `RuntimeError: CUDA error: misaligned address`.

### CODE REVIEW

File: src/rnn_no_discrim.json [hunk 1]
```diff
@@ -1,26 +1,29 @@
...
     "lr_gamma": 1e-2,
     "teacher_init_val": 1,
     "teacher_gamma": 0.98,
-    "teacher_decay_start": 10,
-    "teacher_decay_end": 70,
+    "teacher_decay_start": 50,
+    "teacher_decay_end": 200,
     "grad_clip": 1.0,
-    "batch_size": 16,
+    "train_batch_size": 4,
```
Comment (Lucaskabela) [inline src/rnn_no_discrim.json]: Might be able to try a larger batch size no with the:
`with torch.no_grad():`
in inference!  I can reliably do 16

File: src/train.py [hunk 2]
```diff
@@ -19,16 +20,17 @@
...
+# DEVICE, WRITER are the only global variable
 def adjust_learning_rate(optimizer, e_in, step_num, warmup_step=4000):
     lr = e_in**-0.5 * min(step_num * warmup_step**-1.5, step_num**-0.5)
     for param_group in optimizer.param_groups:
         param_group['lr'] = lr
 
 class BatchGetter():
     def __init__(self, args, supervised_dataset, unsupervised_dataset, full_dataset):
-        self.batch_size = args.batch_size
+        self.batch_size = args.train_batch_size
```
Comment (Lucaskabela) [inline src/train.py:34]: Good insight on splitting this into train/eval batch size!

File: src/train.py [hunk 3]
```diff
@@ -482,6 +618,21 @@ def train(args):
     model.eval()
     return model
 
+
+def log_tb_example(model, ex, step, name="train"):
+    model.eval()
+    if WRITER:
+        ex_mel = torch.tensor(ex["mel"]).unsqueeze(0)
+        ex_text = torch.LongTensor(ex["text"]).unsqueeze(0)
+        text_pred = model.asr(None, ex_mel.to(DEVICE), infer=True).squeeze(dim=0).detach().cpu().numpy()
```
Comment (Lucaskabela) [inline src/train.py]: need to pass lengths here!

File: src/train.py [hunk 4]
```diff
@@ -482,6 +618,21 @@ def train(args):
...
+        speech_pred, speech_stop_pred = model.tts(ex_text.to(DEVICE), None, infer=True)
+        speech_pred = speech_pred.squeeze(dim=0).detach().cpu().numpy()
+        WRITER.add_text(f"{name}/text_gold", sequence_to_text(ex["text"]), step)
```
Comment (Lucaskabela) [inline src/train.py]: We should trim this to length if it is not all ready, same with speech prediction!

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

**Comment by liaojh1998** [comment]
Rebased on top of latest `mel_loss/bill`.

Added:
- Schedulers for changing learning rate for transformers.

Code is tested using train and eval batch size of 8.

Some notes on the `RuntimeError`:
- The error only occurs in `crossmodel_step()`, and specifically, `model.cm_speech_in(mel, mel_len)` every time when trying to do inference on the encoded speech sequence for text.
- The error only occurs on accessing `stop_lens[stop_mask] = i` in text inference. Try printing `stop_lens` or printing `choice` there will result in this error as well.

**Comment by liaojh1998** [comment]
`RuntimeError` example log, but happens during `evaluate()` call as well.
image

**Comment by liaojh1998** [comment]
The RNN optimizations here looks good. Thanks for working on this Lucas!

**Pull request merged**